### ISSUE TITLE
stay logged in for longer than 30 minutes - expire what?

### ISSUE THREAD

**Issue by CoderJason123** [aspnet/Identity#1766]
Hello.  So there's a ton of information regarding cookies and expiration, and the ticket vs the cookie containing the ticket.  I get that.

I just need my users to be able to stay logged in, for let's say for 8 hours, if they choose "remember" me, which then gets passed to signinmanager.signinasync().  My "cookie" shows 2 weeks of life in the browser .. which I believe this is from the options.expiretimespan property.  I'm ok with this.

I just do not understand from other articles and logged issues how to allow that 14 day cookie to actually contain "something" (need the term please) that allows them to close the browser, re-open it in 4 hours, and have to log back in because of a 401.  What is the term i am looking for, and where can I set it to make their "credentials via the cookie" value say for 4 hours.  4 hours is probably a good example.

**Comment by blowdart** [comment]
@HaoK can you explain how this all hangs together, and then @Rick-Anderson and I can get this documented in a friendly form?

**Comment by CoderJason123** [comment]
So let me throw this out there just as a quick recap --- i just want someone to stay "logged in", including if they close their browser.  I'd like them to stay logged in for anywhere from 8-24 hours.  

It's extremely confusing between cookie lifetime, authentication ticket, security validator.

What "pieces" do i need to hook up/ modify / etc, to make this work?  Thank you kindly.

**Comment by brockallen** [comment]
Set the IsPersistent flag on the AuthenticationOptions that you pass into SignInAsync. For example:

https://github.com/IdentityServer/IdentityServer4/blob/dev/src/Host/Quickstart/Account/AccountController.cs#L124

**Comment by HaoK** [comment]
Yeah the overview of all the settings are here: https://docs.microsoft.com/en-us/aspnet/core/security/authentication/cookie?view=aspnetcore-2.1&tabs=aspnetcore2x

But the general idea is if you want the cookie to survive closing the browser, you need to set IsPersistent on the AuthenticationProperties passed to SignIn.

How long the cookie is valid is set by the ExpireTimeSpan.

The security validator won't really affect anything unless you change something that causes identity to invalidate the cookie, which means the next time it checks the cookie will be treated as invalid even if its not expired (i.e. the user changes their password, or wants to sign out everywhere)

**Comment by CoderJason123** [comment]
@HaoK  -- ok -- then i must still be doing something wrong, because I have all of those settings completed.  I am running this on linux, and I also have set up a file store for the machine key, which works great by the way. 

So, the cookie expiretimespan --- is this how long the cookie is valid (mine seems to default to 2 weeks, so i would have assumed i can stay logged in with is persistent = true, for 2 weeks), or as some other articles have pointed out, how long the "authentication" ticket in the cookie is valid?

**Comment by HaoK** [comment]
Its easiest just to show your startup code, its very easy to set the wrong thing that looks like the right thing :)

**Comment by CoderJason123** [comment]
`
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (env.IsDevelopment())
            {
                //app.UseDeveloperExceptionPage();
                app.UseHttpStatusCodeExceptionMiddleware();
                app.UseExceptionHandler();

            }
            else
            {
                app.UseHttpStatusCodeExceptionMiddleware();
                app.UseExceptionHandler();
            }

            app.UseStaticFiles(new StaticFileOptions
            {

            });

            app.UseForwardedHeaders(new ForwardedHeadersOptions
            {
                ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto
            });

            app.UseAuthentication();

            app.UseMvc(routes =>
                routes.MapRoute(
                    name: "default",
                    template: "{controller}/{action}/{id?}",
                    defaults: new { controller = "Home", action = "Index" })

            );
        }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddSingleton<IUserStore<Account>, UserRepository>();
            services.AddSingleton<IRoleStore<ApplicationRole>, RoleRepository>();

            services.AddIdentity<Account, ApplicationRole>(options =>
            // Password settings
            {
                options.Password.RequireDigit = true;
                options.Password.RequiredLength = 8;
                options.Password.RequireNonAlphanumeric = false;
                options.Password.RequireUppercase = true;

                // lockout
                options.Lockout.DefaultLockoutTimeSpan = TimeSpan.FromMinutes(5);
                options.Lockout.MaxFailedAccessAttempts = 5;
                options.Lockout.AllowedForNewUsers = true;

                // Signin settings
                options.SignIn.RequireConfirmedEmail = true;
                options.SignIn.RequireConfirmedPhoneNumber = false;

                // User settings
                options.User.RequireUniqueEmail = true;

            }).AddDefaultTokenProviders();

          

            services.AddAuthentication();

            services.AddMvc().AddJsonOptions(options =>
            {
                options.SerializerSettings.DateTimeZoneHandling = DateTimeZoneHandling.Unspecified;
            });

            services.Configure<SqlSettings>(Configuration.GetSection(nameof(SqlSettings)));
            services.Configure<FileSettings>(Configuration.GetSection(nameof(FileSettings)));
            services.Configure<PaymentSettings>(Configuration.GetSection(nameof(PaymentSettings)));

            var fileConfigSettings = Configuration.GetSection(nameof(FileSettings)).Get<FileSettings>();

            services.AddDataProtection().PersistKeysToFileSystem(new DirectoryInfo(fileConfigSettings.KeyPath));

            services.ConfigureApplicationCookie(options =>
            {
                options.SlidingExpiration = true;

                options.Events.OnSignedIn = cxt =>
                {

                    return Task.CompletedTask;
                };
                options.LoginPath = "/home/login";
                options.Events.OnRedirectToLogin = context =>
                {
                    var path = context.Request.Path.Value.ToLower();
                    if (!path.Contains("/views") && !path.Equals("/settings") && !path.Equals("/contacts"))
                    {
                        context.Response.Clear();
                        context.Response.StatusCode = 401;
                        return Task.FromResult(0);
                    }
                    context.Response.Redirect(context.RedirectUri);
                    return Task.FromResult(0);
                };
            });

            services.AddSingleton(typeof(IConnectionManager), typeof(ConnectionManager));
        }
    }
`

**Comment by HaoK** [comment]
And how about your SignIn call, that's where IsPersistent needs to be specified

**Comment by CoderJason123** [comment]
@HaoK -- yes, that flag is set as well, when the user checks the box of course!  i'm not having a problem with IsPersistent.  I'm having a problem with users being signed out.  Are you saying it is simply a matter of setting expiretimespan and that setting alone (with is persistent set to true), say to a value of TimeSpan.FromHours(8), will provide me a token embedded in the cookie, and that's all I need to do and it should be working?

**Comment by brockallen** [comment]
The one setting on the cookie options is used for both persistent and non-persistent (which is sort of silly since they need to work two different ways). In short, when you set the persistent value you should also set the expiration on the auth options.

**Comment by CoderJason123** [comment]
@brockallen @HaoK  -- here's an update.  I have now set the expire time span property on the cookie options.  I deleted my browser cache and cookies, redeployed, and logged back in.  I notice that the cookie expiration shows as 12-31-1969 23:59:59, as it did before I set this value.  I personally don't care what this date means.

However, I'm going to try hitting a secured url in a few hours, and see if I stay logged in.  I appreciate your help and hope I just missed this one property or misunderstood it as the cookie lifetime.

**Comment by Halofreak1990** [comment]
Another confusing part, is the fact that if your user store does not implement IUserSecurityStampStore<TUser> the SecurityStampValidator will always invalidate the identity cookie after 30 minutes (or whatever timespan you've got configured).

On top of this, the Identity component has its own cookie settings, separate from other cookie-related middleware, leading to a confusing mess, and no easy way to increase the login session timeout.

In my situation, I set up the Identity, specify the password requirements, and add the default token providers, before setting up the authentication middleware and associated cookie. I'd expect _these_ cookie settings to be the ones used, but, even though I specified a custom security stamp validator to bypass the aforementioned issue, I still get logged out after 30 minutes, and from placing a breakpoint, I can see that my custom 'validator' is simply never invoked.

**Comment by brockallen** [comment]
> When i check "remember me", the cookie shows a week from today as the expiration!! (i set expiretimespan from days (7). When i don't check that, it shows 12-31-1969....sound correct?

The expiration is in the [protected] payload of the cookie, not the cookie expiration flag that the browser dev tools show you.
